upcast to most precise dtype within their category before the comparison

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61031 remove UsageError in favor of ValueError
* #60639 change language from array to tensor
* #60638 relax type equality constraint
* #60637 Don't check stride by default
* #60636 Use maximum of tolerances in case of mismatching dtypes
* **#60536 upcast to most precise dtype within their category before the comparison**
* #60254 tests for diagnostics in callable `msg` in `torch.testing.assert_close`
* #60163 update docstring examples of `torch.testing.assert_close`
* #60091 Improve error messages of `torch.testing.assert_close` in case of mismatching values

`torch.isclose` does not do this bool tensors, which results in a test failure since subtraction (`abs(actual - expected)`) is not supported for them (see #58981). Since the `dtype` is already checked at this point, we can safely move the upcasting before `torch.isclose` is invoked.

Differential Revision: [D29556356](https://our.internmc.facebook.com/intern/diff/D29556356)

pmeier committed:
upcast to most precise dtype within their category before the comparison

`torch.isclose` does not do this bool tensors, which results in a test failure since subtraction (`abs(actual - expected)`) is not supported for them (see #58981). Since the `dtype` is already checked at this point, we can safely move the upcasting before `torch.isclose` is invoked.

[ghstack-poisoned]

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -313,7 +312,9 @@ def _check_values_close(
         (Optional[AssertionError]): If check did not pass.
     """
-
+    dtype = torch.float64 if actual.dtype.is_floating_point else torch.int64
```
This would not work for complex tensors, but as long as we are checking the components individually, this is sufficient.

mruberry replied:
How does assertEqual do this?

If we're comparing fp16 and fp64, for example, and the user doesn't specify rtol or atol, which rtol and atol are used? The one for fp16 or the one for fp64?

pmeier replied:
This has nothing to do with the tolerances, since they are selected before we even hit that line. Plus, in `_check_values_close` the inputs are already equalized, so `actual` and `expected` have the same `dtype` here.

---

> How does assertEqual do this?

Internally, `self.assertEqual` calls `self._compareTensors` for the actual comparison. In there we hit 

https://github.com/pytorch/pytorch/blob/00896cb9ed931bb816cf1a5e5f13ddc6b229378b/torch/testing/_internal/common_utils.py#L1267-L1269

in case that we didn't specify the tolerances (recall that only specifying one is not allowed and will fail earlier). The default tolerances are selected by computing the maximum of both input dtypes:

https://github.com/pytorch/pytorch/blob/00896cb9ed931bb816cf1a5e5f13ddc6b229378b/torch/testing/_internal/common_utils.py#L1208-L1226

> If we're comparing fp16 and fp64, for example, and the user doesn't specify rtol or atol, which rtol and atol are used? The one for fp16 or the one for fp64?

The one for fp16. 

This is different what `assert_close` does now as observed in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866654600, but I will send a PR that syncs the behavior.

pmeier replied:
This is resolved in #60636 later in this stack.

pmeier committed:
Update on "upcast to most precise dtype within their category before the comparison"


`torch.isclose` does not do this bool tensors, which results in a test failure since subtraction (`abs(actual - expected)`) is not supported for them (see #58981). Since the `dtype` is already checked at this point, we can safely move the upcasting before `torch.isclose` is invoked.

[ghstack-poisoned]

pmeier committed:
Update on "upcast to most precise dtype within their category before the comparison"


`torch.isclose` does not do this bool tensors, which results in a test failure since subtraction (`abs(actual - expected)`) is not supported for them (see #58981). Since the `dtype` is already checked at this point, we can safely move the upcasting before `torch.isclose` is invoked.

[ghstack-poisoned]

pmeier committed:
Update on "upcast to most precise dtype within their category before the comparison"


`torch.isclose` does not do this bool tensors, which results in a test failure since subtraction (`abs(actual - expected)`) is not supported for them (see #58981). Since the `dtype` is already checked at this point, we can safely move the upcasting before `torch.isclose` is invoked.

[ghstack-poisoned]

mruberry reviewed (approved):
OK

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29556356).
